File: src/web/audio/add_mixer_state_history_record.js
/**
 * Add mixer state history record
 */
function add_mixer_state_history_record(ch_state) {
	// 0x2ee4 = 0x2ef0 in p8 binary
	const history_idx = c[(ch_state + 0x2ee4) >> 2];

	// 0x2024 = 0x2030 in p8 binary
	const sfx_tick = c[(ch_state + 0x2024) >> 2];

	// 0x2ef0 = 0x2f00 in p8 binary
	c[(ch_state + 0x2ef0 + history_idx * 0x18) >> 2] = sfx_tick;

	// step tick
	// 0x2ef0 = 0x2f00 in p8 binary
	// 0x2028 = 0x2034 in p8 binary
	c[(ch_state + 0x2ef0 + history_idx * 0x18 + 4) >> 2] =
		c[(ch_state + 0x2028) >> 2];

	const sfx_ptr = c[(ch_state + 0x2020) >> 2];

	// sfx ptr
	// 0x2ef0 = 0x2f00 in p8 binary
	// 0x2020 = 0x2028 in p8 binary
	c[(ch_state + 0x2ef0 + history_idx * 0x18 + 8) >> 2] = sfx_ptr;

	// step idx
	// 0x2020 = 0x2028 in p8 binary
	if (sfx_ptr) {
		let spd = c[(sfx_ptr + 8) >> 2];
		spd = spd > 1 ? spd : 1;

		// 0x2ef0 = 0x2f00 in p8 binary
		c[(ch_state + 0x2ef0 + history_idx * 0x18 + 0xc) >> 2] = Math.floor(
			sfx_tick / spd,
		);
	} else {
		// 0x2ef0 = 0x2f00 in p8 binary
		c[(ch_state + 0x2ef0 + history_idx * 0x18 + 0xc) >> 2] = 0;
	}

	// pat idx
	// 0x2ef0 = 0x2f00 in p8 binary
	// 0x2d20 = 0x2d2c in p8 binary
	c[(ch_state + 0x2ef0 + history_idx * 0x18 + 0x10) >> 2] =
		c[(ch_state + 0x2d20) >> 2];

	// patterns played
	// 0x2ef0 = 0x2f00 in p8 binary
	c[(ch_state + 0x2ef0 + history_idx * 0x18 + 0x14) >> 2] = c[codo_state_636];

	// 0x2ee4 = 0x2ef0 in p8 binary
	c[(ch_state + 0x2ee4) >> 2] = history_idx + 1;
}

File: src/binary/audio/synth/add_mixer_state_history_record.c
#include "../../globals.h"

/**
 * Add Mixer State History Record
 */
void add_mixer_state_history_record(long ch_state) {
  const int history_idx = *(&ch_state + 0x2ef0);

  const int sfx_tick = *(&ch_state + 0x2030);

  *(&ch_state + 0x2f00 + history_idx * 0x18) = sfx_tick;

  // step tick
  *(&ch_state + 0x2f00 + history_idx * 0x18 + 4) = *(&ch_state + 0x2034);

  const long *sfx_ptr = &ch_state + 0x2028;

  *(&ch_state + 0x2f00 + history_idx * 0x18 + 8) = *sfx_ptr;

  // step idx
  if (*sfx_ptr) {
    int spd = *(sfx_ptr + 8);
    spd = spd > 1 ? spd : 1;

    *(&ch_state + 0x2f00 + history_idx * 0x18 + 0xc) = sfx_tick / spd;
  } else {
    *(&ch_state + 0x2f00 + history_idx * 0x18 + 0xc) = 0;
  }

  // pat idx
  *(&ch_state + 0x2f00 + history_idx * 0x18 + 0x10) = *(&ch_state + 0x2d2c);

  // patterns played
  *(&ch_state + 0x2f00 + history_idx * 0x18 + 0x14) = codo_state[636 >> 2];

  // 0x2ee4 = 0x2ef0 in p8 binary
  *(&ch_state + 0x2ee4) = history_idx + 1;
}
